### PR TITLE
feat: restore wallet from KeyStore JSON file

### LINKED ISSUE #2661: [Demand]  Restore wallet from keystore JSON file
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

Restore the Ethereum wallet from a JSON key store file.

### Steps

+ Step 1: Read a JSON file.
+ Step 2: Validate the content with key store JSON schema. If failed then throw an error.
+ Step 3: Use lib to restore private key from the key store JSON.
+ Step 4: Recover wallet from the private key and a user given password.

### UI

+ Add a new tab - **Keystore** in the `<DashboardWalletImportDialog />`.
+ A multi-line text field or drop down area for the JSON file.
+ A single-line text field for the password.

### Screenshots

image
image

## Notes

+ [keythereum](https://github.com/ethereumjs/keythereum)
+ [ethereumjs/ethereumjs-wallet](https://github.com/ethereumjs/ethereumjs-wallet)

### PR DESCRIPTION
**Pull request by septs**
closes #2661

### CODE REVIEW

**Comment by septs** [comment]
@guanbinrui plz review `packages/maskbook/src/plugins/Wallet/services/keystore/README.md`

**Review by guanbinrui — DISMISSED** [review]

**Comment by septs** [comment]
@AinuX need test

**Review by guanbinrui — APPROVED** [review]

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Base.tsx [hunk 1]
```diff
@@ -244,14 +244,15 @@ interface DashboardDialogWrapperProps {
     size?: 'small' | 'medium'
     content?: React.ReactNode
     footer?: React.ReactNode
+    wrapperClassName?: string
```
Comment (Jack-Works) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Base.tsx:247]: Please follow our pattern on this. Use `Props extends withClasses<'wrapper'>`

**Comment by septs** [comment]
@AinuX ready, plz test

**Review by Jack-Works — APPROVED** [review]

**Review by guanbinrui — DISMISSED** [review]

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet/Import.tsx [hunk 2]
```diff
@@ -211,7 +191,21 @@ export function DashboardWalletImportDialog(props: WrappedDialogProps<object>) {
...
+                        name,
+                        address: addr,
+                        _private_key_: privateKey,
+                    })
+                    break
+                case 2:
                     const words = mnemonic.split(' ')
+                    if (words.length !== 12) {
+                        throw new Error(t('import_failed'))
+                    }
```
Comment (guanbinrui) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet/Import.tsx:208]:
```suggestion
                    if (words.length !== 12) throw new Error(t('import_failed'))
```

**Review by guanbinrui — DISMISSED** [review]

**Review by Jack-Works — DISMISSED** [review]

**Review by Jack-Works — APPROVED** [review]

**Review by guanbinrui — APPROVED** [review]

**Review by Jack-Works — DISMISSED** [review]

File: packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet/Import.tsx [hunk 3]
```diff
@@ -49,43 +51,28 @@ export function DashboardWalletImportDialog(props: WrappedDialogProps<object>) {
 
     const hdWallet = useWalletHD()
 
+    // wallet name
     const [name, setName] = useState('')
```
Comment (UncleBill) [inline packages/maskbook/src/extension/options-page/DashboardDialogs/Wallet/Import.tsx:55]:
```suggestion
    const [walletName, setWalletName] = useState('')
```

nit: I think we don't need comments here, let the variable names doc themself.

**Review by UncleBill — DISMISSED** [review]

**Review by Jack-Works — APPROVED** [review]

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**